### ISSUE TITLE
[User] Crash on 1.4.1 when generating with Typical < 1 and sampler order 6,5,4,3,2,1,0

### ISSUE THREAD

**Issue by tigerbears** [LostRuins/koboldcpp#398]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [X] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [X] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

Using the Lite UI on kcpp 1.4.1, I attempted to generate some text. Sampler parameter settings included Typical set to various values other than 1. (e.g. 0.7) The sampler order was also set to 6,5,4,3,2,1,0. I expected some text to be generated in the UI.

# Current Behavior

Immediately after processing the prompt's tokens, Kobold.cpp crashes with a segmentation fault reported in the shell session and the UI displays an error message. Here's a sample from one test run on my Mac:

`Processing Prompt (1 / 1 tokens)
./startup.sh: line 2: 66508 Segmentation fault: 11  python3 ./koboldcpp.py --gpulayers 256 --ropeconfig 1.0 10000 --contextsize 4096 --noblas --model [redacted for PII]`

Using the API for generation has a similar outcome, but simply outputs "Segmentation fault" to the terminal instead.

# Environment and Context

* I reproduced this with multiple models, first seen with codellama-34b.Q4_K_M.gguf from TheBloke. (Mm, I may need to confirm that one again. Let me know if you can't repro with another model of your choice; I tried 3 and all failed.)
* I reproduced it on Ubuntu 22.04, compiled with LLAMA_OPENBLAS=1 LLAMA_CLBLAST=1 LLAMA_CUBLAS=1 and offloading all layers to a 4090 for inference.
* I reproduced it on MacOS 13.4.1 on a MacBook Pro M2, compiled with LLAMA_METAL=1.
* It was not reproducible on the Mac when running Kobold.cpp 1.4.0
* Launch parameters are included below.
* I can reproduce this on my Ubuntu box whether using the Lite UI directly, or using another front-end to send requests to Kobold.cpp's API.

* I compared a set of known-good-on-1.4.1 settings with the first group of settings I used that resulted in the crash, trying each one that was different until I narrowed it down to the combination of the sampler order (6,5,4,3,2,1,0) and Typical being set to some value other than 1. (I used 0.7 most recently.) *Both* must be set that way to trigger the crash. For example, switching to the recommended sampler order produces expected output. I didn't try many other sampler orders. Here is a complete set of input parameters that fails for me:

`Input: {"n": 1, "max_context_length": 1024, "max_length": 100, "rep_pen": 1.1, "temperature": 0.7, "top_p": 0.92, "top_k": 0, "top_a": 0, "typical": 0.7, "tfs": 1, "rep_pen_range": 320, "rep_pen_slope": 0.7, "sampler_order": [6, 5, 4, 3, 2, 1, 0], "prompt": "Kobold.cpp is awesome", "quiet": true}`

* I specifically confirmed that those settings don't crash on 1.4.0 on my Mac.
* I don't recall why that sampler order and typical value were set to these values. If there's something nonsensical about the combination, cool. I suspect I had been randomly experimenting at one point and left it that way. (I usually use Kobold via API with the baseline sampler order.)

* Physical (or virtual) hardware you are using, e.g. for Linux:

Let me know if you can't repro this on your system and I can provide these; I'm running out of time for the moment!

# Steps to Reproduce

* Launch Kobold.cpp 1.4.1 with any model. GGUF or GGML models both appear to have this behavior.
* Sample Ubuntu launch parameters:
* `CUDA_VISIBLE_DEVICES=0 python3 koboldcpp.py --model [redacted for PII] --host 0.0.0.0 --threads 1 --blasbatchsize 1024 --gpulayers 100 --contextsize 4096 --ropeconfig 1.0 82000 --usecublas normal 0 mmq`
* Sample Mac launch parameters: 
* `python3 ./koboldcpp.py --gpulayers 256 --ropeconfig 1.0 10000 --contextsize 4096 --noblas --model [redacted for PII]]`
* Configure the samplers to include **both Typical 0.7 and a sampler order of 6,5,4,3,2,1,0**
* Attempt to generate text with any prompt.

# Failure Logs

Here's a snippet from the raw crash log on my Mac. Sorry it's not symbolicated (or whatever equivalent applies here); I also couldn't get core files on my Ubuntu box for some reason, but hopefully this should help get a little closer at least.

Thread 4 Crashed:
0   koboldcpp_default.so          	       0x1033515c0 0x10325c000 + 1004992
1   koboldcpp_default.so          	       0x1032e8ea8 llama_sample_typical + 892
2   koboldcpp_default.so          	       0x103320a64 SampleLogits(float const*, int, int, int, float, float, float, float, float, float, float, std::__1::mersenne_twister_engine<unsigned int, 32ul, 624ul, 397ul, 31ul, 2567483615u, 11ul, 4294967295u, 7ul, 2636928640u, 15ul, 4022730752u, 18ul, 1812433253u>&, int, float, float, std::__1::vector<samplers, std::__1::allocator<samplers>> const&) + 904
3   koboldcpp_default.so          	       0x1033243d0 gpttype_generate(generation_inputs, generation_outputs&) + 6496
4   koboldcpp_default.so          	       0x1032c9a4c generate + 68
5   libffi.dylib                  	       0x19a058050 ffi_call_SYSV + 80
6   libffi.dylib                  	       0x19a060af8 ffi_call_int + 1208
7   libffi.dylib                  	       0x19a060af8 ffi_call_int + 1208
8   libffi.dylib                  	       0x19a060af8 ffi_call_int + 1208

**Comment by LostRuins** [comment]
Did you do a `make clean` followed by a `make`? 

Also I noticed you are using the `koboldcpp_default.so` library, while you are attempting to use the GPU. If compiled for LLAMA_CUBLAS=1 for example the loaded library would be `-koboldcpp_cublas.so` so I believe something is wrong there.

If that is not the case, could you see if you have similar problems upstream llama.cpp

**Comment by kex243** [comment]
same error on complied non cuda version for 'godlike' order of samplers.

**Comment by tigerbears** [comment]
> Did you do a `make clean` followed by a `make`?

I'm pretty sure I did on Ubuntu. To be sure, I just did a fresh clone in a new directory and was able to reproduce the problem. (I did it twice really; one time running `make clean` first just to rule that out.) 

On the Mac, my installation was also a fresh clone in a new directory, so I didn't run `make clean` first.

> Also I noticed you are using the `koboldcpp_default.so` library, while you are attempting to use the GPU. If compiled for LLAMA_CUBLAS=1 for example the loaded library would be `-koboldcpp_cublas.so` so I believe something is wrong there.
> 
> If that is not the case, could you see if you have similar problems upstream llama.cpp

That stack trace is from running Kobold.cpp on my Mac, which I compiled with only the LLAMA_METAL=1 flag. I also launch it with --noblas, but I'm not clear if that flag is relevant.

I've been unable to get clear debugging data when kcpp is installed and running on my Ubuntu box, though I experience the same crash. 

I could eventually get through building upstream, but looking through the docs, I don't see a way to apply the sampler order which is a key component of the issue. If I'm just missing it somewhere, I'd appreciate a pointer for when I'm able to try it at some point.

What happens when you do a generation with that problematic pair of sampler settings?

In retesting for this note, I also found that the sampler order 5,6,4,3,2,1,0 (paired with typical @ 0.7) resulted in a crash.

**Comment by LostRuins** [comment]
Hi, sorry for the inconvenience, this should be fixed in v1.42. Can check if it's still causing issues?

**Comment by dburm** [comment]
`$ python3 koboldcpp/koboldcpp.py --threads 8 --model models/airoboros-l2-13b-2.1.Q5_K_M.gguf --contextsize 3072
***
Welcome to KoboldCpp - Version 1.42
Warning: OpenBLAS library file not found. Non-BLAS library will be used.
Initializing dynamic library: koboldcpp_default.so
==========
Namespace(model='models/airoboros-l2-13b-2.1.Q5_K_M.gguf', model_param='models/airoboros-l2-13b-2.1.Q5_K_M.gguf', port=5001, port_param=5001, host='', launch=False, lora=None, config=None, threads=8, blasthreads=8, psutil_set_threads=False, highpriority=False, contextsize=3072, blasbatchsize=512, ropeconfig=[0.0, 10000.0], stream=False, smartcontext=False, unbantokens=False, bantokens=None, usemirostat=None, forceversion=0, nommap=False, usemlock=False, noavx2=False, debugmode=0, skiplauncher=False, hordeconfig=None, noblas=False, useclblast=None, usecublas=None, gpulayers=0, tensor_split=None)
==========
Loading model: ./models/airoboros-l2-13b-2.1.Q5_K_M.gguf 
[Threads: 8, BlasThreads: 8, SmartContext: False]

---
Identified as LLAMA model: (ver 6)
Attempting to Load...
---
Segmentation fault: 11
`
Macbook Pro M1. Segfaults for contextsize greater than 2048

**Comment by LostRuins** [comment]
There was a bug in 1.42, I released a fix in 1.42.1, can you try that instead?

**Comment by tigerbears** [comment]
> Hi, sorry for the inconvenience, this should be fixed in v1.42. Can check if it's still causing issues?

Cool! I saw the discussion on Discord and was glad more details were found. I rebuilt from the concedo_experimental branch yesterday and confirmed that the problem with that sampler order and typical < 1 was resolved there. To be sure, I'll check again from the main branch / 1.42.1 release when I have a chance, probably later today, but I think you squashed the bug. I actually don't use those settings, so no inconvenience here, but am grateful for the improvement all the same. Thank you!

**Comment by dburm** [comment]
Seems works now.

Probably another issue: ggml/gguf q5 failed to run with `gpulayers` and `contextsize` > 2048:
```llama_new_context_with_model: kv self size  = 2400.00 MB
llama_new_context_with_model: compute buffer total size =  291.47 MB
llama_new_context_with_model: max tensor size =   128.17 MB
GGML_ASSERT: ggml-metal.m:1185: false
./run.sh: line 33: 64526 Abort trap: 6           python3 koboldcpp/koboldcpp.py --contextsize 3072 --gpulayers 10 --threads 8 --smartcontext --skiplauncher --model ./${MODEL}
```
q4 looks working in this configuration

**Comment by LostRuins** [comment]
I'm not sure if metal has support for q5. Try using accelerate.

**Comment by dburm** [comment]
Seems so. It works without LLAMA_METAL but has noticeable performance degradation. 
However metat+gpulayers+contextsize 2048 works well

**Comment by tigerbears** [comment]
OK, I did some more testing and am pretty sure the original bug has been fixed in 1.42.1. I can generate plenty with the Godlike preset which has both typical < 1 and 6,5,4,3,2,1,0 as the sampler sequence. Nice!

However, it's unclear if the fix is complete, or if there's a separate and rarer issue that was hidden behind it. I can still generate crashes for really short prompts ("hey there yo") if I drop Godlike's temperature down to 0.51. This is using mythomax-l2-13b.ggmlv3.q5_K_M with the startup parameters above. I don't know if it manifests with other models. If there's more in context than that (let's call it "unlikely") prompt, it seems to be able to proceed.

At first this made it look like the original issue wasn't fixed after all, but after trying stock Godlike and other (sometimes trash) test prompts, it seems the more important bug is squashed. My gut says this is a separate issue, but I don't know what tokens are coming back in this case, or what's actually causing the segfault, to be more helpful than that at this time. Up to you if you want it spilt off or use this issue for that. Let me know if I can do anything else to help.

**Comment by LostRuins** [comment]
So the problem was that in GGUF, the logits were swapped from positive to negative values. This means that settings a token logit to 0 to be banned (such as EOS) no longer worked. To fix this, I originally set the logit to negative infinity. However, that caused problems with the typical sampler e^-inf returns 0, and log(0) is infinity. The fix sets the ban to the lowest logit value -8 instead, which works well enough.
